OutOfSchool/OutOfSchool.BusinessLogic/Services/TagService.cs

```diff
@@ -0,0 +1,145 @@
+    /// <param name="repository">Repository.</param>
+    /// <param name="logger">Logger.</param>
+    /// <param name="localizer">Localizer.</param>
+    /// <param name="mapper">Mapper.</param>
+    public TagService(
+        IEntityRepository<long, Tag> repository,
+        ILogger<TagService> logger,
+        IStringLocalizer<SharedResource> localizer,
+        IMapper mapper)
+    {
+        this.localizer = localizer ?? throw new ArgumentNullException(nameof(localizer));
+        this.repository = repository ?? throw new ArgumentNullException(nameof(repository));
+        this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        this.mapper = mapper ?? throw new ArgumentNullException(nameof(mapper));
+    }
+
+    /// <inheritdoc/>
+    public async Task<IEnumerable<TagDto>> GetAll(LocalizationType localization = LocalizationType.Ua)
+    {
+        logger.LogInformation($"Getting all Tags, {localization} localization, started.");
+
+        var tags = await repository.GetAll().ConfigureAwait(false);
+
+        logger.LogInformation(!tags.Any()
+            ? "SocialGroup table is empty."
+            : $"All {tags.Count()} records were successfully received from the Tag table");
+
+        return tags.Select(x =>
+        new TagDto()
+        {
+            Id = x.Id,
+            Name = localization == LocalizationType.En ? x.NameEn : x.Name,
+        }).ToList();
+    }
+
+    /// <inheritdoc/>
+    public async Task<TagDto> GetById(long id, LocalizationType localization = LocalizationType.Ua)
+    {
+        logger.LogInformation($"Getting Tag by Id, {localization} localization, started. Looking Id = {id}.");
+
+        var tag = await repository.GetById(id).ConfigureAwait(false);
+
+        if (tag == null)
+        {
+            throw new ArgumentOutOfRangeException(
+                nameof(id),
+                localizer["The id cannot be greater than number of table entities."]);
+        }
+
+        logger.LogInformation($"Successfully got a SocialGroup with Id = {id} and {localization} localization.");
+
+        return new TagDto()
+        {
+            Id = tag.Id,
+            Name = localization == LocalizationType.En ? tag.NameEn : tag.Name,
+        };
+    }
+
+    /// <inheritdoc/>
+    public async Task<TagCreate> Create(TagCreate dto)
+    {
+        logger.LogInformation("Tag creating was started.");
+
+        var tag = mapper.Map<Tag>(dto);
+
+        var newTag = await repository.Create(tag).ConfigureAwait(false);
+
+        logger.LogInformation($"Tag with Id = {newTag?.Id} created successfully.");
+
+        return mapper.Map<TagCreate>(newTag);
+    }
+
+    /// <inheritdoc/>
+    public async Task<TagDto> Update(TagDto dto, LocalizationType localization = LocalizationType.Ua)
+    {
+        logger.LogInformation($"Updating Tag with Id = {dto?.Id}, {localization} localization, started.");
+
+        var tagLocalized = await repository.GetById(dto.Id).ConfigureAwait(false);
```

Comment:
(Inconsistent nullability) In two lines you have both `dto?.Id` and `dto.Id`. That's wrong. 

- Either assert that `dto != null` if you worry about it, 
- or trust compiler type checks (I hope we have nullability feature switched on for this file)

But for any case please don't use that elvis operator (?.) for `dto`. Same thing for `tag`

Reply:
Please write a note here clarifiing your plans to perform any changes or not.
At the moment I see this comment open, heving no response and no changes of related code (consistent working out of the nullables) implemented

Reply:
done here